x/storage/keeper/keeper.go

```diff
@@ -2323,3 +2416,162 @@ func (k Keeper) SetTag(ctx sdk.Context, operator sdk.AccAddress, grn types2.GRN,
 
 	return nil
 }
+
+func (k Keeper) UpdateObjectContent(
+	ctx sdk.Context, operator sdk.AccAddress, bucketName, objectName string, payloadSize uint64,
+	opts types.UpdateObjectOptions,
+) error {
+	store := ctx.KVStore(k.storeKey)
+
+	bucketInfo, found := k.GetBucketInfo(ctx, bucketName)
+	if !found {
+		return types.ErrNoSuchBucket
+	}
+	err := bucketInfo.CheckBucketStatus()
+	if err != nil {
+		return err
+	}
+	objectInfo, found := k.GetObjectInfo(ctx, bucketName, objectName)
+	if !found {
+		return types.ErrNoSuchObject
+	}
+	// check object status
+	if objectInfo.ObjectStatus != types.OBJECT_STATUS_SEALED {
+		return types.ErrUpdateObjectNotAllowed.Wrapf("The object is not sealed yet")
+	}
+	if objectInfo.IsUpdating {
+		return types.ErrObjectIsUpdating.Wrapf("The object is already being updated")
+	}
+	// check permission
+	effect := k.VerifyObjectPermission(ctx, bucketInfo, objectInfo, operator, permtypes.ACTION_UPDATE_OBJECT_CONTENT)
+	if effect != permtypes.EFFECT_ALLOW {
+		return types.ErrAccessDenied.Wrapf(
+			"The operator(%s) has no updateObjectContent permission of the bucket(%s), object(%s)",
+			operator.String(), bucketName, objectName)
+	}
+
+	// check payload size
+	if payloadSize > k.MaxPayloadSize(ctx) {
+		return types.ErrTooLargeObject
+	}
+
+	// primary sp
+	sp := k.MustGetPrimarySPForBucket(ctx, bucketInfo)
+	// a sp is not in service, neither in maintenance
+	if sp.Status != sptypes.STATUS_IN_SERVICE && !k.fromSpMaintenanceAcct(sp, operator) {
+		return errors.Wrap(types.ErrNoSuchStorageProvider, "the storage provider is not in service")
+	}
+
+	nextVersion := objectInfo.Version + 1
+
+	if payloadSize == 0 {
+		internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+		err := k.UnChargeObjectStoreFee(ctx, bucketInfo, k.MustGetInternalBucketInfo(ctx, bucketInfo.Id), objectInfo)
+		if err != nil {
+			return err
+		}
+		k.SetInternalBucketInfo(ctx, bucketInfo.Id, internalBucketInfo)
+		err = k.DeleteObjectFromVirtualGroup(ctx, bucketInfo, objectInfo)
+		if err != nil {
+			return err
+		}
+
+		objectInfo.UpdatedAt = ctx.BlockTime().Unix()
+		objectInfo.Version = nextVersion
+		objectInfo.PayloadSize = 0
+		objectInfo.Checksums = opts.Checksums
+		objectInfo.UpdatedBy = operator.String()
+		objectInfo.ContentType = opts.ContentType
+
+		_, err = k.SealEmptyObjectOnVirtualGroup(ctx, bucketInfo, objectInfo)
+		if err != nil {
+			return err
+		}
+	} else {
+		objectInfo.IsUpdating = true
```

Comment:
`IsUpdating` is a temp field. It seems better we just put into `ShadowObjectInfo` - so it will be deleted with shadow object to save storage.